sdk/nodejs/runtime/invoke.ts

```diff
@@ -81,22 +130,173 @@ export async function invoke(tok: string, props: Inputs, opts?: InvokeOptions):
             })), label);
 
         // If there were failures, propagate them.
-        const failures: any = resp.getFailuresList();
-        if (failures && failures.length) {
-            let reasons = "";
-            for (let i = 0; i < failures.length; i++) {
-                if (reasons !== "") {
-                    reasons += "; ";
-                }
-                reasons += `${failures[i].getReason()} (${failures[i].getProperty()})`;
-            }
-            throw new Error(`Invoke of '${tok}' failed: ${reasons}`);
-        }
+        processPotentialFailures(tok, resp);
 
         // Finally propagate any other properties that were given to us as outputs.
         return deserializeProperties(resp.getReturn());
     }
     finally {
         done();
     }
+
+    async function getProviderRefAsync() {
+        const provider = getProvider(tok, opts);
+
+        if (ProviderRef.isInstance(provider)) {
+            return provider;
+        }
+        else if (Resource.isInstance(provider)) {
+            return await ProviderRef.get(provider);
+        }
+        else {
+            return undefined;
+        }
+    }
+}
+
+function invokeSync(tok: string, props: any, opts: InvokeOptions, syncInvokes: SyncInvokes): Promise<any> {
+    const label = `Invoking function: tok=${tok} synchronously`;
+    log.debug(label + (excessiveDebugOutput ? `, props=${JSON.stringify(props)}` : ``));
+
+    const serialized = serializePropertiesSync(props);
+    log.debug(`Invoke RPC prepared: tok=${tok}` + excessiveDebugOutput ? `, obj=${JSON.stringify(serialized)}` : ``);
+
+    const providerRef = getProviderRefSync();
+    const req = createInvokeRequest(tok, serialized, providerRef, opts);
+
+    // Encode the request.
+    const reqBytes = Buffer.from(req.serializeBinary());
+
+    // Write the request length.
+    const reqLen = Buffer.alloc(4);
+    reqLen.writeUInt32BE(reqBytes.length, /*offset:*/ 0);
+    fs.writeSync(syncInvokes.requests, reqLen);
+    fs.writeSync(syncInvokes.requests, reqBytes);
+
+    // Read the response.
+    const respLenBytes = Buffer.alloc(4);
+    fs.readSync(syncInvokes.responses, respLenBytes, /*offset:*/ 0, /*length:*/ 4, /*position:*/ null);
+    const respLen = respLenBytes.readUInt32BE(/*offset:*/ 0);
+    const respBytes = Buffer.alloc(respLen);
+    fs.readSync(syncInvokes.responses, respBytes, /*offset:*/ 0, /*length:*/ respLen, /*position:*/ null);
+
+    // Decode the response.
+    const resp = providerproto.InvokeResponse.deserializeBinary(new Uint8Array(respBytes));
+
+    // If there were failures, propagate them.
+    processPotentialFailures(tok, resp);
```

Comment:
nit: `checkFailures` feels better here. Clarifying that `propagate` means `throw` would be nice, too.

Reply:
sure.